### PR TITLE
[DNM][RFC] RPMsg-Lite and OpenAMP side-by-side

### PR DESCRIPTION
**Pull request by skordal**
This pull request adds an example application using OpenAMP to do multiprocessor communication. It is adapted from the recently posted RPMsg-Lite example application, from PR #5960. This application is intended to provide a starting-point for comparison and benchmarking of RPMsg-Lite and OpenAMP, see issue #3065.

The relevant commit in this PR is the most recent commit. However, since the code is dependent on changes included in PR #5960, the commits from that PR is also included here.

I have added build instructions in a README.md file, located in samples/subsys/ipc/openamp/README.rst.

Some details:
- The OpenAMP example uses the SRAMX RAM block as shared memory, to provide better separation between shared RAM and the RAM that is being used by the application. This also simplified the linking of the application.
- Heap size had to be overridden in the default board configuration for both the cores in the LPC54114. This is because the heap size is needed (at compile-time) for libmetal, and when compiling libraries outside of the zephyr application folders, only the default board configuration is used.
- Due to the use of static libraries, the order that the libraries are provided to the linker in matters. The libkernel.a must be linked last, and the CMakeLists.txt files contains code to fix this.

### CODE REVIEW

**Comment by carlescufi** [comment]
CC @stanislav-poboril @wjliang

**Review by dbkinder — COMMENTED** [review]
Doc looks OK, but there are branch conflicts.

File: samples/subsys/ipc/openamp/README.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+OpenAMP Example Application
```
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: This file should be called README.rst (with an .rst extension)

Add a line before this with a document label to make it easier to reference externaly:

    .. _openAMP_example:

File: samples/subsys/ipc/openamp/README.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+===========================
```
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: Our Zephyr doc standards use ### underlines for H1 headings, *** for H2, === for H3, and --- for H4.
So use ### underlines for this heading

File: samples/subsys/ipc/openamp/README.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+Overview
+--------
```
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: use *** underline for this heading

File: samples/subsys/ipc/openamp/README.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+This application demostrates how to use OpenAMP with Zephyr. It is designed to
+be an OpenAMP counterpart to the RPMsg-Lite example for the LPCXpresso5411x
+chip, to be used for comparison and benchmarking of the two RPC libraries.
+
+Building this application
+-------------------------
```
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: use *** underlines here

File: samples/subsys/ipc/openamp/README.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+Building and running this application requires that libmetal and OpenAMP is
```
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: change that last ``is`` to ``are``

File: samples/subsys/ipc/openamp/README.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+compiled first. Clone the libmetal and OpenAMP repositories in an appropriate
+location:
```
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: maybe add:

    location (outside the Zephyr code tree):

File: samples/subsys/ipc/openamp/README.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    $ cd build-master
+    $ cmake -DWITH_ZEPHYR=ON -DBOARD=lpcxpresso54114 ..
+    $ make
+    $ cd ../build-remote
+    $ cmake -DWITH_ZEPHYR=ON -DBOARD=lpcxpresso54114_m0 ..
+    $ make
+
+Compile and build OpenAMP for both cores:
+
+    $ cd open-amp
```
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: would this be instead:  ``cd ../../open-amp``

File: samples/subsys/ipc/openamp/README.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    $ cd ../build-remote
+    $ cmake -DWITH_ZEPHYR=ON -DWITH_PROXY=OFF -DBOARD=lpcxpresso54114_m0 \
+        -DLIBMETAL_INCLUDE_DIR=<path to libmetal>/build-remote/lib/include \
+        -DLIBMETAL_LIB=<path to libmetal>/build-remote/lib/libmetal.a ..
+    $ make
+
+Compile the remote application, by going to the toplevel folder of the example
+(where this file is located) and running the following commands:
+
+    $ cd remote
```
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: I think the directory you want them to go to is ``cd $ZEPHYR_BASE/samples/subsys/ipc/ipm_mcux/remote``

File: samples/subsys/ipc/openamp/README.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        -DLIBMETAL_INCLUDE_DIR=<path to libmetal>/build-master/lib/include \
+        -DLIBMETAL_LIB=<path to libmetal>/build-master/lib/libmetal.a ..
+    $ make
+    $ cd ../build-remote
+    $ cmake -DWITH_ZEPHYR=ON -DWITH_PROXY=OFF -DBOARD=lpcxpresso54114_m0 \
+        -DLIBMETAL_INCLUDE_DIR=<path to libmetal>/build-remote/lib/include \
+        -DLIBMETAL_LIB=<path to libmetal>/build-remote/lib/libmetal.a ..
+    $ make
+
+Compile the remote application, by going to the toplevel folder of the example
```
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: I think you can delete the ", by going to the..." part of this sentence by the cd change below

File: samples/subsys/ipc/openamp/README.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    $ cd remote
+    $ mkdir build
+    $ cmake -DBOARD=lpcxpresso54114_m0 \
+        -DLIBMETAL_INCLUDE_DIR=<path to libmetal>/build-remote/lib/include \
+        -DLIBMETAL_LIBRARY=<path to libmetal>/build-remote/lib/libmetal.a \
+        -DOPENAMP_INCLUDE_DIR=<path to openamp>/lib/include \
+        -DOPENAMP_LIBRARY=<path to openamp>/build-remote/lib/libopen_amp.a ..
+    $ make
+
+Compile the master application, by going to the toplevel folder of the example
```
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: Same here (deleting the " by going to the...")

File: samples/subsys/ipc/openamp/README.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+and running the following commands:
+    $ cd master
```
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: and I suspect this should be ``cd $ZEPHYR_BASE/samples/subsys/ipc/ipm_mcux/master``

File: samples/subsys/ipc/openamp/README.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+be an OpenAMP counterpart to the RPMsg-Lite example for the LPCXpresso5411x
+chip, to be used for comparison and benchmarking of the two RPC libraries.
+
+Building this application
+-------------------------
+
+Building and running this application requires that libmetal and OpenAMP is
+compiled first. Clone the libmetal and OpenAMP repositories in an appropriate
+location:
+
```
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: Use the code-block directive to get this formatted correctly:

    .. code-block:: bash

       git clone https://github.com/OpenAMP/libmetal.git
       git clone https://github.com/OpenAMP/open-amp.git

File: samples/subsys/ipc/openamp/README.md [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    $ git clone https://github.com/OpenAMP/libmetal.git
+    $ git clone https://github.com/OpenAMP/open-amp.git
+
+Compile and build libmetal for both cores:
+
+    $ cd libmetal
```
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: use a code-block here too:

    .. code-block:: bash

       mkdir build-master build-remote
       cd build-master
       cmake -DWITH_ZEPHYR=ON -DBOARD=lpcxpresso54114 ..
       make
       cd ../build-remote
       cmake -DWITH_ZEPHYR=ON -DBOARD=lpcxpresso54114_m0 ..
       make

File: samples/subsys/ipc/openamp/README.md (hunk 7 above)
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: use code-block here too

File: samples/subsys/ipc/openamp/README.md (hunk 8 above)
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: code-block

File: samples/subsys/ipc/openamp/README.md (hunk 11 above)
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: you know what to do (code-block)

File: samples/subsys/ipc/openamp/README.md [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    $ mkdir build
+    $ cmake -DBOARD=lpcxpresso54114 \
+        -DLIBMETAL_INCLUDE_DIR=<path to libmetal>/build-master/lib/include \
+        -DLIBMETAL_LIBRARY=<path to libmetal>/build-master/lib/libmetal.a \
+        -DOPENAMP_INCLUDE_DIR=<path to openamp>/lib/include \
+        -DOPENAMP_LIBRARY=<path to openamp>/build-master/lib/libopen_amp.a ..
+    $ make
+
+Flash the project to the board from the master build directory:
+    $ make flash
```
Comment (dbkinder) [inline samples/subsys/ipc/openamp/README.md]: one more code-block

**Review by dbkinder — CHANGES_REQUESTED** [review]
.rst file for our Zephyr docs

**Review by dbkinder — APPROVED** [review]
Takk!  (But you still have merge conflicts that need to be resolved)

**Comment by skordal** [comment]
Tenative comparison of executable sizes based on a build with default settings and GCC 7.2.1. The size of the image for the second core has been subtracted from the sizes of the `.text` section for the master applications.

| Executable                | .text | .data | .bss  |
|:--------------------------|------:|------:|------:|
| RPMsg-Lite master         | 16782 |   328 | 15744 |
| RPMsg-Lite remote         | 11040 |   128 | 8544  |
| OpenAMP master            | 24000 |   664 | 9512  |
| OpenAMP remote            | 18106 |   464 | 8456  |

**Comment by MarekNovakNXP** [comment]
Cannot we have both implementations in /ext? Is it not what is /ext intended for?
Why there are for example two crypto libraries (tinycrypt and mbedtls) supported? I can see a similar situation here... 
Can we discuss on this?

**Comment by PetrLukas** [comment]
@skordal could you please post also map files, you have used for code size comparison? 

Did RPMSG-lite use static or dynamic allocation option? Default settings uses dynamic allocation option. There could be also used static to decrease codesize quite a bit.

Thanks, Petr

**Comment by skordal** [comment]
@PetrLukas I added the mapfiles to my comment above. I will get some numbers for RPMsg-Lite using the static API and add them to the table.

**Comment by PetrLukas** [comment]
@skordal Thank you for analysis and map files. 

I was trying to find what is reason for big difference in master BSS size. It seems that in case of RPMSG Lite shared memory buffers are included in statistics (rpmsg_sh_mem), while in case of openAMP implementation it is not. If it would be included it would add another 30kB.

> #define SHM_START_ADDRESS       0x04000400
> #define SHM_SIZE                0x7c00

Speed comparison - I think it more or less depends on placement of shared buffers. In case of OpenAMP they are placed in SRAMX. It may impact performance for big buffer transfers. We should pace shared buffers to the same SRAM for both implementations.

**Comment by skordal** [comment]
Thanks, I did not notice that the shared memory was included in the .bss section for the RPMsg-Lite code. I have updated the numbers to account for this.

I also did an experiment with the shared memory placement. Overall, there seems to be an average performance drop of around 1 % when using SRAMX. I added the numbers to the table in the comment above for comparison.

**Comment by MarekNovakNXP** [comment]
Hello,
I reviewed the figures concerning the code size provided by @skordal  and I tried to remove the zephyr code from the comparison and to keep only RPMsg-Lite  / OpenAMP RPMsg code.
The difference in code size is about 60%.

I attach the excell sheet, where the computation is done. Let me know if you find any error in this benchmark, which reflects more the reality about code size comparison between OpenAMP RPMsg and RPMsg-Lite.

B/CODE | remote | master |   | remote | master
-- | -- | -- | -- | -- | --
lite | 3106 | 3256 |   | -60.8224 | -58.9304
openamp | 8174 | 7928 |   | 3.102926 | REF



[comparison.xlsx](https://github.com/zephyrproject-rtos/zephyr/files/1774971/comparison.xlsx)

**Comment by carlescufi** [comment]
@skordal @MaureenHelm should we close this now?

**Comment by skordal** [comment]
That's fine by me, there should be a new PR with a new example when OpenAMP is ready to be included in Zephyr.

**Pull request closed without merging**